oxidation/crates/parsec_core_fs/src/storage/chunk_storage.rs

```diff
@@ -0,0 +1,404 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+
+use diesel::dsl::count_star;
+use diesel::{sql_query, ExpressionMethods, Insertable, QueryDsl, RunQueryDsl};
+use std::sync::Mutex;
+
+use parsec_api_crypto::SecretKey;
+use parsec_api_types::{ChunkID, DateTime, DEFAULT_BLOCK_SIZE};
+
+use super::local_database::{SqliteConn, SQLITE_MAX_VARIABLE_NUMBER};
+use crate::error::{FSError, FSResult};
+use crate::extensions::coalesce_total_size;
+use crate::schema::chunks;
+
+#[derive(Insertable, AsChangeset)]
+#[diesel(table_name = chunks)]
+struct NewChunk<'a> {
+    pub chunk_id: &'a [u8],
+    pub size: i32,
+    pub offline: bool,
+    pub accessed_on: Option<f32>,
+    pub data: &'a [u8],
+}
+
+pub(crate) trait ChunkStorageTrait {
+    fn conn(&self) -> &Mutex<SqliteConn>;
+    fn local_symkey(&self) -> &SecretKey;
+
+    fn create_db(&self) -> FSResult<()> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        sql_query(
+            "CREATE TABLE IF NOT EXISTS chunks (
+                chunk_id BLOB PRIMARY KEY NOT NULL, -- UUID
+                size INTEGER NOT NULL,
+                offline BOOLEAN NOT NULL,
+                accessed_on REAL, -- Timestamp
+                data BLOB NOT NULL
+            );",
+        )
+        .execute(conn)
+        .map_err(|_| FSError::CreateTable("chunks"))?;
+        Ok(())
+    }
+
+    #[cfg(test)]
+    fn drop_db(&self) -> FSResult<()> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        sql_query("DROP TABLE IF EXISTS chunks;")
+            .execute(conn)
+            .map_err(|_| FSError::DropTable("chunks"))?;
+        Ok(())
+    }
+
+    // Size and chunks
+
+    fn get_nb_blocks(&self) -> FSResult<i64> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(count_star())
+            .first(conn)
+            .map_err(|_| FSError::QueryTable("chunks: get_nb_blocks"))
+    }
+
+    fn get_total_size(&self) -> FSResult<i64> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(coalesce_total_size())
+            .first(conn)
+            .map_err(|_| FSError::QueryTable("chunks: get_total_size"))
+    }
+
+    fn is_chunk(&self, chunk_id: ChunkID) -> FSResult<bool> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(count_star())
+            .filter(chunks::chunk_id.eq((*chunk_id).as_ref()))
+            .first::<i64>(conn)
+            .map_err(|_| FSError::QueryTable("chunks: is_chunk"))
```

Comment:
could be useful to save the error returned by `first::<i64>(conn)` for debug purpose